File: go/raindrops/raindrops.go
package raindrops

import "fmt"
import "strings"

func Convert(n int) string {
	out := make([]string, 0)
	if n%3 == 0 {
		out = append(out, "Pling")
	}
	if n%5 == 0 {
		out = append(out, "Plang")
	}
	if n%7 == 0 {
		out = append(out, "Plong")
	}
	if len(out) == 0 {
		return fmt.Sprintf("%d", n)
	}
	return strings.Join(out, "")
}

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

import qualified Data.Map as M

factorWords n =
  M.elems (M.filterWithKey (\k _ -> n `mod` k == 0) words)
  where 
    words = M.fromList [ (3, "Pling"), (5, "Plang"), (7, "Plong") ]

convert :: Int -> String
convert n = 
  case factorWords n of 
    [] -> show n
    ws -> foldl1 (++) ws
